### DISCUSSION TITLE
How to "Refresh" Vertex Quality Array Between Calls

### DISCUSSION THREAD

**Discussion by zoyt** [cnr-isti-vclab/PyMeshLab#202] [Q&A] [answered]
Hello,

I've been setting up a function to apply a Poisson sampling across a mesh, and then record all the faces that are within a given geometric distance from each of the Poisson samples. I've got each of those components working individually, but I've run into issues integrating the two. I figure there must be some way to do this easily, as this feels like a pretty common issue to run into, but I haven't been able to find anything in the documentation. 

Here's a simplified version of the code:

```
from numpy import full, int32, where, asarray, unique, full, any, array, in1d
import pymeshlab as ml
from itertools import compress

poissonDiskRad = 5
maxGeodesicDist = 1.3235

ms = ml.MeshSet()  # OZ: create a meshset object for Meshlab to process

ms.load_new_mesh('subdivided_file.stl') # test mesh shortcut to avoid long remeshing step

def sortedClosestFacesToPoint(coords):

    def sorting(vert):
        return ms.mesh(0).vertex_quality_array()[vert]

    ms.apply_filter("colorize_by_geodesic_distance_from_a_given_point", startpoint=[coords[0],coords[1],coords[2]],
                    maxdistance=maxGeodesicDist) # stores the verts within the search cutoff in quality array
    pointsWithinDist = where(ms.mesh(0).vertex_quality_array() > 0)[0]

    vertsWithinDist = asarray(sorted(pointsWithinDist, reverse=True, key=sorting)) # return the vertices within
    # cutoff in ascending distance

    face_matrix = array(ms.mesh(0).face_matrix())
    flat = face_matrix.flatten()

    def fastFacesFromVerts(vertList):
        inVertList = in1d(flat, vertList)
        inFlattenedRows = asarray(list(compress(range(len(inVertList)), inVertList)))
        return unique(inFlattenedRows//3) # convert out of flattened matrix, isolate unique faces

    return fastFacesFromVerts(vertsWithinDist)
```
This code runs perfectly fine with a call such as: 
`sortedClosestFacesToPoint([60.913090, 0.231804, -42.673534])
`
However, when I try to integrate the Poisson Disk Sampling, something goes wrong. The Poisson code is as follows: 
```
ms.apply_filter('poisson_disk_sampling', samplenum=1000, radius=poissonDiskRad, approximategeodesicdistance=True)
poissonSamples = ms.mesh(1).vertex_matrix()
```

To summarize the interesting/problematic behavior:
```
#(prior to running the poisson code)
sortedClosestFacesToPoint(coords_A) -> returns: faces_A
sortedClosestFacesToPoint(coords_B) -> returns: faces_B
#(run poisson code)
sortedClosestFacesToPoint(coords_A) -> returns: faces_B
```

It seems to me that something about running the Poisson code impacts the mesh layer that the code is applied on, like in desktop Meshlab, but I am not certain how to get around this problem. I would love to hear your thoughts. 

Thank you very much!

Owen

**Comment by zoyt** [comment, marked as answer]
In case anyone else runs into this problem, I seem to have solved this using the `ms.set_current_mesh(0)` function, which I was apparently just using incorrectly before when I tried to use it. I simply dropped it in as the last line on my Poisson sampling code chunk.
